contracts/modules/TokenExchanger.sol

```diff
@@ -33,139 +35,238 @@ contract TokenExchanger is OnlyOwnerModule {
+     * source token into a given (minimum) amount of a given destination token. The path is a sequence of sets of weighted routes where the destination
+     * token of a set of weighted routes matches the source token of the next set of weighted routes in the path.
+     * @param _mintPrice gasPrice (in wei) at the time the gas tokens were minte by ParaSwap. 0 means gas token will not be used by ParaSwap
      */
-    function trade(
+    function multiSwap(
         address _wallet,
         address _srcToken,
+        address _destToken,
         uint256 _srcAmount,
+        uint256 _minDestAmount,
+        uint256 _expectedDestAmount,
+        IAugustusSwapper.Path[] calldata _path,
+        uint256 _mintPrice
+    )
+        external
+        onlyWalletOwner(_wallet)
+        onlyWhenUnlocked(_wallet)
+    {
+        // Verify that the exchange adapters used have been authorised
+        verifyExchangeAdapters(_path);
+        // Approve source amount if required
+        uint previousAllowance = approveToken(_wallet, _srcToken, _srcAmount);
+        // Perform trade and emit event
+        doMultiSwap(
+            _wallet,
+            _srcToken,
+            _destToken,
+            _srcAmount,
+            _minDestAmount,
+            _expectedDestAmount,
+            _path,
+            _mintPrice);
+        // Restore the previous allowance if needed. This should only be needed when the previous allowance
+        // was infinite. In other cases, paraswap.multiSwap() should have used exactly the additional allowance
+        // granted to it and therefore the previous allowance should have been restored.
+        restoreAllowance(_wallet, _srcToken, previousAllowance);
+    }
+
+    /**
+     * @dev Lets the owner of the wallet execute a "buy" trade (fixed destination amount, variable source amount).
+     * @param _wallet The target wallet
+     * @param _srcToken The address of the source token.
+     * @param _destToken The address of the destination token.
+     * @param _maxSrcAmount The maximum amount of source tokens to use for the trade.
+     * @param _destAmount The exact amount of destination tokens to buy.
+     * @param _expectedDestAmount The expected amount of destination tokens (used only in ParaSwap's events).
+     * @param _routes Set of weighted ParaSwap routes. Each route specifies an exchange to use to convert a given (maximum) amount of a given
+     * source token into a given (exact) amount of a given destination token.
+     * @param _mintPrice gasPrice (in wei) at the time the gas tokens were minte by ParaSwap. 0 means gas token will not be used by ParaSwap
+     */
+    function buy(
+        address _wallet,
+        address _srcToken,
         address _destToken,
-        uint256 _maxDestAmount,
-        uint256 _minConversionRate
+        uint256 _maxSrcAmount,
+        uint256 _destAmount,
+        uint256 _expectedDestAmount,
```

Comment:
Is `_expectedDestAmount` returned by the API and is it different than `_destAmount` in a buy? Maybe `_destAmount` is not needed?

Reply:
This is a typo: it should be `_expectedSrcAmount`. It's given by the API and is only used by Paraswap's `Bought` event.

Reply:
Fixed